precovery/frame_db.py

```diff
@@ -561,32 +582,43 @@ def load_hdf5(
                 data_offset=offset,
                 data_length=length,
             )
-            self.idx.add_frame(frame)
+            frames_to_add.append(frame)
+
+        self.idx.add_frames(frames_to_add)
 
     def _open_data_files(self):
-        matcher = os.path.join(self.data_root, "frames*.data")
-        files = glob.glob(matcher)
+        matcher = os.path.join(self.data_root, "**/frames*.data")
+        files = sorted(glob.glob(matcher, recursive=True))
         for f in files:
             abspath = os.path.abspath(f)
             name = os.path.basename(f)
-            self.data_files[name] = open(abspath, "rb")
-        self.n_data_files = len(self.data_files) - 1
-        if self.n_data_files <= 0:
-            self.new_data_file()
-
-    def _current_data_file_name(self):
-        return "frames_{:08d}.data".format(self.n_data_files)
+            year_month_str = os.path.basename(os.path.dirname(abspath))
+            dataset_id = os.path.basename(os.path.dirname(os.path.dirname(abspath)))
+            data_uri = f"{dataset_id}/{year_month_str}/{name}"
+            if dataset_id not in self.n_data_files.keys():
+                self.n_data_files[dataset_id] = {}
+            if year_month_str not in self.n_data_files[dataset_id].keys():
+                self.n_data_files[dataset_id][year_month_str] = 0
+            self.data_files[data_uri] = open(abspath, "rb")
+            self.n_data_files[dataset_id][year_month_str] += 1
+
+    def _current_data_file_name(self, dataset_id: str, year_month_str: str):
+        return "{}/{}/frames_{:08d}.data".format(
+            dataset_id, year_month_str, self.n_data_files[dataset_id][year_month_str]
+        )
 
-    def _current_data_file_full(self):
+    def _current_data_file_full(self, dataset_id: str, year_month_str: str):
         return os.path.abspath(
-            os.path.join(self.data_root, self._current_data_file_name())
+            os.path.join(
+                self.data_root, self._current_data_file_name(dataset_id, year_month_str)
+            )
         )
 
-    def _current_data_file_size(self):
-        return os.stat(self._current_data_file_name()).st_size
+    def _current_data_file_size(self, dataset_id: str, year_month_str: str):
+        return os.stat(self._current_data_file_name(dataset_id, year_month_str)).st_size
```

Comment:
It's pretty strange how the `dataset_id` and `year_month_str` are getting threaded through so many functions, and always as a pair. Is one ever provided separate from the other? I think this is evidence for a `PartitionIdentifier` as a dedicated concept.